src/levels.rs

```diff
@@ -1,5 +1,1053 @@
+                targets.base_level = i;
+            }
+            db_size /= self.opts.level_size_multiplier as u64;
+        }
+
+        let mut tsz = self.opts.base_level_size;
+
+        // Use mem_table_size for Level 0. Because at Level 0, we stop
+        // compactions based on the number of tables, not the size of
+        // the level. So, having a 1:1 size ratio between memtable size
+        // and the size of L0 files is better than churning out 32 files
+        // per memtable (assuming 64MB mem_table_size and 2MB base_table_size).
+        for i in 0..self.levels.len() {
+            if i == 0 {
+                targets.file_size[i] = self.opts.mem_table_size;
+            } else if i <= targets.base_level {
+                targets.file_size[i] = tsz;
+            } else {
+                tsz *= self.opts.table_size_multiplier as u64;
+                targets.file_size[i] = tsz;
+            }
+        }
+
+        targets
+    }
+
+    /// Returns [`LevelHandler`] of bottom level.
+    fn last_level(&self) -> &Arc<RwLock<LevelHandler>> {
+        self.levels.last().unwrap()
+    }
+
+    /// Determines which level to compact.
+    fn pick_compact_levels(&self) -> Vec<CompactionPriority> {
+        let targets = self.level_targets();
+        let mut prios = vec![];
+
+        let mut add_priority = |level, score| {
+            let pri = CompactionPriority {
+                level,
+                score,
+                adjusted: score,
+                targets: targets.clone(),
+                drop_prefixes: vec![],
+            };
+            prios.push(pri);
+        };
+
+        // Add L0 priority based on the number of tables.
+        add_priority(
+            0,
+            self.levels[0].read().unwrap().num_tables() as f64
+                / self.opts.num_level_zero_tables as f64,
+        );
+
+        for i in 1..self.levels.len() {
+            // We must ensure lock order, by only obtain one lock at a time.
+            let del_size = {
+                let cpt_status = self.cpt_status.read().unwrap();
+                cpt_status.levels[i].del_size
+            };
+            let level = self.levels[i].read().unwrap();
+            // There could be inconsistency data which causes `size < 0`.
+            // We may safely ignore this situation.
+            // TODO: Check if we could make it more stable.
+            let size = if del_size <= level.total_size {
+                level.total_size - del_size
+            } else {
+                0
+            };
+            add_priority(i, size as f64 / targets.target_size[i] as f64);
+        }
+
+        assert_eq!(prios.len(), self.levels.len());
+
+        // TODO: Adjust score.
+
+        prios.pop(); // remove last level
+        let mut x: Vec<CompactionPriority> = prios.into_iter().filter(|x| x.score > 1.0).collect();
+        x.sort_by(|x, y| x.adjusted.partial_cmp(&y.adjusted).unwrap());
+        x.reverse();
```

Comment:
Don't need reverse if change the line above to `|x, y| y.adjusted.partial_cmp(&x.adjusted).unwrap()`

Reply:
Yes, that's a better method.